File: soluciones/src/main/java/repositorio/Repositorio.java
package repositorio;

import java.util.List;

// https://medium.com/@pererikbergman/repository-design-pattern-e28c0f3e4a30

// Repositorio para entidades gestionadas con id

public interface Repositorio <T, K> {
    
    K add(T entity) throws RepositorioException;
    
    void update(T entity) throws RepositorioException, EntidadNoEncontrada;
    
    void delete(T entity) throws RepositorioException, EntidadNoEncontrada;

    T getById(K id) throws RepositorioException, EntidadNoEncontrada;
    
	List<T> getAll() throws RepositorioException;

	List<K> getIds();
}

File: bookle.NET/Models/Repositorio.cs

using System;
using System.Collections.Generic;


namespace Repositorio
{
    public interface Repositorio<T, K>
    {

        K add(T entity);

        void update(T entity);

        void delete(T entity);

        T getById(K id);

        List<T> getAll();

        List<K> getIds();
    }

    public class EntidadNoEncontrada : SystemException
    {

        public EntidadNoEncontrada(String msg, Exception causa) : base(msg, causa)
        {

        }

        public EntidadNoEncontrada(String msg) : base(msg)
        {

        }
    }

    public class RepositorioException : SystemException
    {

        public RepositorioException(String msg, Exception causa) : base(msg, causa)
        {

        }

        public RepositorioException(String msg) : base(msg)
        {

        }

    }

}